src/main/java/com/yelp/nrtsearch/server/grpc/LuceneServer.java

```diff
@@ -633,22 +635,27 @@ private String onCompletedForIndex(String indexName) {
                   "onCompleted, addDocumentRequestQueue: %s", addDocumentRequestQueue.size()));
           try {
             // index the left over docs
+            List<Long> gens = new ArrayList<>();
             if (!addDocumentRequestQueue.isEmpty()) {
               logger.debug(
                   String.format(
                       "indexing left over addDocumentRequestQueue of size: %s",
                       addDocumentRequestQueue.size()));
               List<AddDocumentRequest> addDocRequestList = new ArrayList<>(addDocumentRequestQueue);
-              Future<Long> future =
-                  globalState.submitIndexingTask(
-                      new DocumentIndexer(globalState, addDocRequestList));
-              futures.add(future);
+              // Since we are already running in the indexing threadpool run the indexing job
+              // for remaining documents directly. This serializes indexing remaining documents for
+              // multiple indices but avoids deadlocking if there aren't more threads than the
+              // maximum
+              // number of parallel addDocuments calls.
+              long gen = new DocumentIndexer(globalState, addDocRequestList).runIndexingJob();
+              gens.add(gen);
             }
             // collect futures, block if needed
             PriorityQueue<Long> pq = new PriorityQueue<>(Collections.reverseOrder());
+            gens.forEach(pq::offer);
```

Comment:
Do we even need to use a priority queue here? All we care about is the highest gen, we could keep track of that directly.